src/mpi/comm/comm_impl.c

```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
     goto fn_exit;
 }
 
+int MPIR_Comm_create_from_group_impl(MPIR_Group * group_ptr, const char *stringtag,
+                                     MPIR_Info * info_ptr, MPIR_Errhandler * errhan_ptr,
+                                     MPIR_Comm ** p_newcom_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    /* This implementation assumes an internal MPI_COMM_WORLD already exist.
+     *
+     * The next implementation will relax but assume the first call to this function will
+     * either from a "world" pset or "self" pset, during former a comm world will be created.
+     *
+     * The final implementation will further relax and allow partial initializations.
+     */
+
+    /* use macro from uthash.h to calculate hash value */
+    unsigned hash;
+    int n = strlen(stringtag);
+    HASH_VALUE(stringtag, n, hash);
+
+    int tag = hash % (MPIR_Process.attrs.tag_ub);
+
+    if (MPIR_Process.comm_world) {
+        if (!MPIR_Process.comm_world->local_group) {
+            /* need create local_group for MPIR_Comm_create_group to work */
+            MPIR_Group *world_group_ptr;
+            mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world, &world_group_ptr);
+            MPIR_ERR_CHECK(mpi_errno);
+            MPIR_Group_release(world_group_ptr);
+        }
+        MPIR_Comm_create_group(MPIR_Process.comm_world, group_ptr, tag, p_newcom_ptr);
+    } else if (group_ptr->pset_name && strcmp(group_ptr->pset_name, "mpi://WORLD") == 0) {
+        /* TODO: create and dup MPIR_Process.comm_world */
+        MPIR_Assert(0 && "not implemented");
+        goto fn_fail;
+    } else if (group_ptr->pset_name && strcmp(group_ptr->pset_name, "mpi://SELF") == 0) {
```

Comment:
The current code should always fall to `MPIR_Comm_create_group` since we always have the world comm created during `MPI_Session_init`. When user specify `mpi://SELF`, it'll be just group of 1 and it should work.

Reply:
This PR will be kept as assuming `MPI_Session_init` will always be called collectively because this is the only way to keep backward compatibility for the device layer. I'll do separate PR to implement the local MPI_Session_init, at which point, other branches here will become valid possibilities (and will be implemented).